File: ios/Sources/KlinikAPI/TravelAPI.swift
import Foundation
import KlinikCore

/// The travel around an operation (spec M19).
public struct TravelPlan: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    public let patientId: String
    public let arrivalFlight: String?
    public let arrivalAt: Date?
    public let departureFlight: String?
    public let departureAt: Date?
    public let hotelName: String?
    public let hotelAddress: String?
    public let hotelCheckIn: Date?
    public let hotelCheckOut: Date?
    public let greeterName: String?
    public let greeterPhone: String?
    public let transferNote: String?
    public let interpreterName: String?
    public let interpreterLanguage: String?
    public let interpreterPhone: String?
    /// Set by a clinician. Never computed from the surgery date — how long
    /// somebody must wait before flying is a judgement about them, not a sum.
    public let clearedToFlyAt: Date?
    public let notes: String?
    public let version: Int

    public var hasFlights: Bool { arrivalAt != nil || departureAt != nil }
    public var hasHotel: Bool { hotelName != nil }
    public var hasInterpreter: Bool { interpreterName != nil }
    public var isClearedToFly: Bool { clearedToFlyAt != nil }

    /// Whether anything has been filled in at all.
    public var isEmpty: Bool {
        !hasFlights && !hasHotel && !hasInterpreter && greeterName == nil
            && (transferNote?.isEmpty ?? true) && (notes?.isEmpty ?? true)
    }
}

public struct TravelPlanView: Decodable, Sendable, Equatable {
    public let plan: TravelPlan?
    public let clearedToFlyBy: String?
}

/// A change to the trip. Only what somebody filled in is sent.
public struct TravelPlanEdit: Encodable, Sendable, Equatable {
    public var arrivalFlight: String?
    public var arrivalAt: Date?
    public var departureFlight: String?
    public var departureAt: Date?
    public var hotelName: String?
    public var hotelAddress: String?
    public var hotelCheckIn: Date?
    public var hotelCheckOut: Date?
    public var greeterName: String?
    public var greeterPhone: String?
    public var transferNote: String?
    public var interpreterName: String?
    public var interpreterLanguage: String?
    public var interpreterPhone: String?
    public var notes: String?
    public var expectedVersion: Int?

    public init(
        arrivalFlight: String? = nil,
        arrivalAt: Date? = nil,
        departureFlight: String? = nil,
        departureAt: Date? = nil,
        hotelName: String? = nil,
        hotelAddress: String? = nil,
        hotelCheckIn: Date? = nil,
        hotelCheckOut: Date? = nil,
        greeterName: String? = nil,
        greeterPhone: String? = nil,
        transferNote: String? = nil,
        interpreterName: String? = nil,
        interpreterLanguage: String? = nil,
        interpreterPhone: String? = nil,
        notes: String? = nil,
        expectedVersion: Int? = nil
    ) {
        self.arrivalFlight = arrivalFlight
        self.arrivalAt = arrivalAt
        self.departureFlight = departureFlight
        self.departureAt = departureAt
        self.hotelName = hotelName
        self.hotelAddress = hotelAddress
        self.hotelCheckIn = hotelCheckIn
        self.hotelCheckOut = hotelCheckOut
        self.greeterName = greeterName
        self.greeterPhone = greeterPhone
        self.transferNote = transferNote
        self.interpreterName = interpreterName
        self.interpreterLanguage = interpreterLanguage
        self.interpreterPhone = interpreterPhone
        self.notes = notes
        self.expectedVersion = expectedVersion
    }
}

public struct TravelAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    public func forPatient(_ patientId: String) async throws -> TravelPlanView {
        try await client.send(
            Endpoint(method: .get, path: "patients/\(patientId)/travel"),
            as: TravelPlanView.self
        )
    }

    public func mine() async throws -> TravelPlanView {
        try await client.send(Endpoint(method: .get, path: "me/travel"), as: TravelPlanView.self)
    }

    public func save(patientId: String, _ edit: TravelPlanEdit) async throws -> TravelPlan {
        try await client.send(
            Endpoint(
                method: .put,
                path: "patients/\(patientId)/travel",
                body: try JSONEncoder.klinik.encode(edit)
            ),
            as: TravelPlan.self
        )
    }

    /// Needs `medical.decide`: a coordinator may move a hotel booking and may
    /// not decide somebody is fit to fly.
    public func setClearedToFly(patientId: String, cleared: Bool) async throws -> TravelPlan {
        try await client.send(
            Endpoint(
                method: .patch,
                path: "patients/\(patientId)/travel/cleared-to-fly",
                body: try JSONEncoder.klinik.encode(ClearBody(cleared: cleared))
            ),
            as: TravelPlan.self
        )
    }

    private struct ClearBody: Encodable {
        let cleared: Bool
    }
}

File: android/core/network/src/main/kotlin/xyz/klinik/network/TravelApi.kt
package xyz.klinik.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Getting the patient here and home again (spec M14).
 *
 * A health-tourism clinic's file is half clinical and half logistical: a
 * flight, a hotel, somebody meeting them at the airport, an interpreter. None
 * of it is medical, and one field on it is: `clearedToFlyAt` is set by a
 * clinician and never computed from the surgery date, because "seven days
 * after an operation" is a rule of thumb and a signature is a decision.
 */
@Serializable
data class TravelPlan(
    val id: String,
    val patientId: String,
    val arrivalFlight: String? = null,
    val arrivalAt: String? = null,
    val departureFlight: String? = null,
    val departureAt: String? = null,
    val hotelName: String? = null,
    val hotelAddress: String? = null,
    val hotelCheckIn: String? = null,
    val hotelCheckOut: String? = null,
    val greeterName: String? = null,
    val greeterPhone: String? = null,
    val transferNote: String? = null,
    val interpreterName: String? = null,
    val interpreterLanguage: String? = null,
    val interpreterPhone: String? = null,
    /** Set by a clinician; never computed from the surgery date. */
    val clearedToFlyAt: String? = null,
    val notes: String? = null,
    /** Sent back on a save, so two coordinators cannot overwrite each other. */
    val version: Int = 0,
) {
    val isClearedToFly: Boolean get() = clearedToFlyAt != null

    /** Whether anybody has filled any of it in. */
    val isEmpty: Boolean
        get() = listOf(
            arrivalFlight, arrivalAt, departureFlight, departureAt,
            hotelName, greeterName, interpreterName, notes,
        ).all { it.isNullOrBlank() }
}

@Serializable
data class TravelPlanView(
    val plan: TravelPlan? = null,
    /** Who signed off the flight. Null when nobody has. */
    val clearedToFlyBy: String? = null,
)

@Serializable
data class UpsertTravelPlan(
    val arrivalFlight: String? = null,
    val arrivalAt: String? = null,
    val departureFlight: String? = null,
    val departureAt: String? = null,
    val hotelName: String? = null,
    val hotelAddress: String? = null,
    val hotelCheckIn: String? = null,
    val hotelCheckOut: String? = null,
    val greeterName: String? = null,
    val greeterPhone: String? = null,
    val transferNote: String? = null,
    val interpreterName: String? = null,
    val interpreterLanguage: String? = null,
    val interpreterPhone: String? = null,
    val notes: String? = null,
    /**
     * What the editor last read.
     *
     * Two coordinators arranging one patient's travel is the ordinary case in
     * this clinic, and the second save winning silently would lose the first's
     * flight number.
     */
    val expectedVersion: Int? = null,
)

@Serializable
private data class ClearToFlyBody(val cleared: Boolean)

class TravelApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    /** The patient's own copy. Read-only: they do not arrange their own transfer. */
    suspend fun mine(): TravelPlanView =
        decode(client.send(Endpoint(HttpMethod.GET, "me/travel")))

    suspend fun forPatient(patientId: String): TravelPlanView =
        decode(client.send(Endpoint(HttpMethod.GET, "patients/$patientId/travel")))

    suspend fun save(patientId: String, plan: UpsertTravelPlan): TravelPlanView =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.PUT,
                    "patients/$patientId/travel",
                    body = json.encodeToString(UpsertTravelPlan.serializer(), plan),
                ),
            ),
        )

    /**
     * A clinician's signature on the flight home.
     *
     * Separate from the rest of the plan on purpose: everything else on this
     * screen is logistics a coordinator owns, and this one is a medical
     * decision with a name against it.
     */
    suspend fun setClearedToFly(patientId: String, cleared: Boolean): TravelPlanView =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.PATCH,
                    "patients/$patientId/travel/cleared-to-fly",
                    body = json.encodeToString(ClearToFlyBody.serializer(), ClearToFlyBody(cleared)),
                ),
            ),
        )

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }
}
